### ISSUE TITLE
Types Should Not Expose Implementation

### ISSUE THREAD

**Issue by andrewcooke** [JuliaCollections/DataStructures.jl#96]
hi.  obviously this is just a suggestion, and i am very grateful for this library, but it seems to me that things like Stack and Queue should have types that show their contents, not their implementation.  the end user, reading the code, "wants" to see what the stack contains, not how it is implemented.

so, without thinking through any consequences, i'd suggest something more like:

```
immutable Stack{S}
    stack::UnderlyingType{S}
end
```

rather than the `stack{S}` you have at the moment.

**Comment by kmsquire** [comment]
Seems reasonable.  Care to make a pull request?

Cc: @lindahua

**Comment by andrewcooke** [comment]
while i am doing it, can i also change the constructors so that they are more similar to the types used in Base?  for example, instead of

```
Stack(MyType)
```

it would be

```
Stack{MyType}()
```

which seems to be a more normal approach (but wasn't possible when the stack was actually of the internal implementation type).

**Comment by kmsquire** [comment]
Go for it!

**Comment by andrewcooke** [comment]
that last comment about the constructor was just wrong, sorry.  but i've created a pull request that hides the type.  hope it's ok - first time i've actually gone through this whole procedure with julia + github.

**Comment by kmsquire** [comment]
Something like this should work for your constructor change:

```julia
type Stack{T}
    store::Deque{T}
    
    Stack() = new(Deque{T}())
    Stack(blksize::Integer) = new(Deque{T}(blksize))
end

Stack() = Stack{Any}()
Stack(blksize::Integer) = Stack{Any}(blksize)
```

Testing:
```julia
julia> using DataStructures

julia> Stack{Int}()
DataStructures.Stack{Int64}(Deque [Int64[]])

julia> x = Stack{Int}(2048)
DataStructures.Stack{Int64}(Deque [Int64[]])

julia> x.store.blksize
2048
```

(`show()` should probably be overridden to clean up the output.)

**Comment by oxinabox** [comment]
Closed in #164
